lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -1729,6 +1734,167 @@ void NodeOp::getCanonicalizationPatterns(RewritePatternSet &results,
   results.insert<FoldNodeName>(context);
 }
 
+SmallVector<Value> getCompleteWrite(Operation *lhs) {
+  auto lhsTy = lhs->getResult(0).getType();
+  if (!lhsTy.isa<BundleType>() && !lhsTy.isa<FVectorType>())
+    return {};
+
+  DenseMap<uint32_t, Value> fields;
+  for (Operation *user : lhs->getResult(0).getUsers()) {
+    if (user->getParentOp() != lhs->getParentOp())
+      return {};
+    if (auto aConnect = dyn_cast<StrictConnectOp>(user)) {
+      if (aConnect.getDest() == lhs->getResult(0))
+        return {};
+    } else if (auto subField = dyn_cast<SubfieldOp>(user)) {
+      for (Operation *subuser : subField.getResult().getUsers()) {
+        if (auto aConnect = dyn_cast<StrictConnectOp>(subuser)) {
+          if (aConnect.getDest() == subField) {
+            if (fields.count(subField.getFieldIndex())) // duplicate write
+              return {};
```

Comment:
Nice! Thanks for checking for multiple assignment. I was worried that this canonicalizer would be a foot-gun if it did the wrong thing before `ExpandWhens`. 👍